### DISCUSSION TITLE
Error opening Administrator with 4.3.0 or later — exception from setRhinoLanguageVersion

### DISCUSSION THREAD

**Discussion by AaronKelley** [nextgenhealthcare/connect#6132] [Q&A]
Trying to upgrade an ancient instance of Mirth Connect.  It was running 3.10.0.  I went straight to the latest 4.5.0.  The server started up fine but I was unable to connect using the administrator tool, it gave me the error "There was an error connecting to the server at the specified address. Please verify that the server is up and running."  (The server *was* up and running, and I could hit the web site on port 8080/8443 where you can download the admin tool.)

So, I backed up to 3.10.0, which involved a database restore as well to undo the schema upgrade.

I then went forward one version at a time to figure out where it breaks.  I got up to 4.2.0 working, and 4.3.0 breaks.

Here's the details on the admin tool error:

The Mirth Connect Administrator tool tries to go through the login process after I enter my credentials.  It gets up to "Loading extensions...", sits there for a few minutes, and then comes back to the credential entry screen with the error (above) that shows right below the username and password fields.

![image](https://github.com/nextgenhealthcare/connect/assets/404829/9d226220-49c8-4986-9a03-4f13f535c125)

I opened the Java console and it shows a stack trace at the same time that the error appears.

![image](https://github.com/nextgenhealthcare/connect/assets/404829/732c16f6-9ac7-4bfc-92c1-faa0c5ef9d63)

```
java.lang.ExceptionInInitializerError
	at com.mirth.connect.util.JavaScriptSharedUtil.setRhinoLanguageVersion(JavaScriptSharedUtil.java:65)
	at com.mirth.connect.client.ui.Frame.setupFrame(Frame.java:646)
	at com.mirth.connect.client.ui.Mirth.<init>(Mirth.java:66)
	at com.mirth.connect.client.ui.LoginPanel$8.handleSuccess(LoginPanel.java:557)
	at com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:446)
	at com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:417)
	at java.desktop/javax.swing.SwingWorker$1.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.desktop/javax.swing.SwingWorker.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.SecurityException: class "org.mozilla.javascript.NativeBoolean"'s signer information does not match signer information of other classes in the same package
	at java.base/java.lang.ClassLoader.checkCerts(Unknown Source)
	at java.base/java.lang.ClassLoader.preDefineClass(Unknown Source)
	at java.base/java.lang.ClassLoader.defineClass(Unknown Source)
	at java.base/java.security.SecureClassLoader.defineClass(Unknown Source)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(Unknown Source)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(Unknown Source)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(Unknown Source)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at org.mozilla.javascript.Context.<clinit>(Context.java:363)
	... 12 more
```

(The stuff from SLF4J is irrelevant, I think; it shows up well before the error appears and it is even there on my "working" instances.)

I looked in the logs on the server side and there is no indication of trouble.  Mirth Connect appears to start normally and channels seem to be working fine.  I just can't get into the admin tool.

Any ideas?  I can't find *anywhere* where someone is complaining about an exception involving "setRhinoLanguageVersion".

[Edit]
Just noticed the bit further down in the exception stack trace which seems to be the crux of the problem.
```Caused by: java.lang.SecurityException: class "org.mozilla.javascript.NativeBoolean"'s signer information does not match signer information of other classes in the same package```

**Comment by pacmano1** [comment]
did you do a diff and merge on the mirth.properties file?

**Comment by ab-mg-23** [comment]
It sounds like the upgrade isn't replacing something correctly or is leaving an old version of the rhino JAR on the file system. The code signing certificate used to sign the JARs changed in 4.3.0. Check that `client-lib/mirth-client-core.jar` and `client-lib/rhino-1.7.13.jar` are both getting updated and there isn't maybe a `client-lib/rhino-1.7.XX.jar` dangling behind after the upgrade to 4.3.0+.

**Comment by AaronKelley** [comment]
Actually, I was just poking around some more...
This is a Mirth instance that I "inherited" so I don't know why some of this stuff is the way that it is!

I found an old file in the custom-lib folder named "tika-app-1.12.jar" which does have some org.mozilla classes inside, so I am wondering if it is creating a conflict (by not being signed with the same cert as other Mirth JARs).  I'm going to see if I can just remove it without causing a problem, I don't think that we would be actually using it for anything.

I will also check on the JARs that you mentioned @ab-mg-23, thank you.

[Edit]
I can already see that there are both rhino-1.7.12.jar and rhino-1.17.13.jar's sitting here on my 4.2.0 install, so maybe it is that simple...

**Comment by AaronKelley** [comment]
OK, I noticed that there are many duplicate libraries (with different version numbers) in the client-lib folder.  This is from a 4.2.0 install.  Should I clean this up?  I guess I can do a "clean" 4.2.0 install and see what is "supposed" to be there.

```
C:\Program Files\Mirth Connect\client-lib>dir
 Volume in drive C has no label.
 Volume Serial Number is 086B-7F4D

 Directory of C:\Program Files\Mirth Connect\client-lib

03/15/2024  08:15 AM    <DIR>          .
03/15/2024  08:15 AM    <DIR>          ..
11/29/2022  06:48 PM           164,564 autocomplete-2.5.4.jar
11/29/2022  06:48 PM         1,032,703 bcpkix-jdk18on-171.jar
11/29/2022  06:48 PM         6,410,245 bcprov-ext-jdk18on-171.jar
11/29/2022  06:48 PM         6,336,128 bcprov-jdk18on-171.jar
11/29/2022  06:48 PM           698,587 bcutil-jdk18on-171.jar
11/29/2022  06:48 PM           265,666 commons-beanutils-1.9.3.jar
11/29/2022  06:48 PM           373,881 commons-codec-1.13.jar
11/29/2022  06:48 PM           814,781 commons-collections4-4.4.jar
11/29/2022  06:48 PM           627,598 commons-compress-1.17.jar
11/29/2022  06:48 PM           670,020 commons-configuration2-2.7.jar
11/29/2022  06:48 PM           233,569 commons-io-2.6.jar
11/29/2022  06:48 PM           305,721 commons-lang-2.6.jar
11/29/2022  06:48 PM           541,137 commons-lang3-3.9.jar
11/29/2022  06:48 PM            72,172 commons-logging-1.2.jar
11/29/2022  06:48 PM           262,181 commons-text-1.10.0.jar
08/24/2022  01:39 PM           228,685 commons-text-1.8.jar
11/29/2022  06:48 PM           475,588 commons-vfs2-2.1.jar
11/29/2022  06:48 PM           121,268 donkey-model.jar
11/29/2022  06:48 PM         2,975,720 guava-28.2-jre.jar
11/29/2022  06:48 PM           725,914 hapi-base-2.3.jar
11/29/2022  06:48 PM           193,505 hapi-structures-v21-2.3.jar
11/29/2022  06:48 PM           438,817 hapi-structures-v22-2.3.jar
11/29/2022  06:48 PM         2,128,564 hapi-structures-v23-2.3.jar
11/29/2022  06:48 PM         1,354,568 hapi-structures-v231-2.3.jar
11/29/2022  06:48 PM         1,557,003 hapi-structures-v24-2.3.jar
11/29/2022  06:48 PM         2,009,330 hapi-structures-v25-2.3.jar
11/29/2022  06:48 PM         2,001,674 hapi-structures-v251-2.3.jar
11/29/2022  06:48 PM         2,211,445 hapi-structures-v26-2.3.jar
11/29/2022  06:48 PM         2,355,709 hapi-structures-v27-2.3.jar
11/29/2022  06:48 PM         2,483,289 hapi-structures-v28-2.3.jar
11/29/2022  06:48 PM         2,546,897 hapi-structures-v281-2.3.jar
11/29/2022  06:48 PM           197,087 hk2-api-2.4.0-b31.jar
11/29/2022  06:48 PM           196,893 hk2-locator-2.4.0-b31.jar
11/29/2022  06:48 PM           115,767 hk2-utils-2.4.0-b31.jar
11/05/2020  05:57 PM           826,222 httpclient-4.5.11.jar
11/29/2022  06:48 PM           830,173 httpclient-4.5.13.jar
11/29/2022  06:48 PM           358,395 httpcore-4.4.13.jar
11/05/2020  05:57 PM            50,075 httpmime-4.5.11.jar
11/29/2022  06:48 PM            51,430 httpmime-4.5.13.jar
11/29/2022  06:48 PM            34,803 istack-commons-runtime-3.0.6.jar
11/29/2022  06:48 PM            80,799 jackson-annotations-2.11.3.jar
11/05/2020  05:57 PM            78,727 jackson-annotations-2.9.10.jar
11/29/2022  06:48 PM           368,516 jackson-core-2.11.3.jar
11/05/2020  05:57 PM           341,604 jackson-core-2.9.10.jar
11/29/2022  06:48 PM         1,483,922 jackson-databind-2.11.3.jar
11/05/2020  05:57 PM         1,412,253 jackson-databind-2.9.10.jar
11/29/2022  06:48 PM           978,545 jai_imageio.jar
11/29/2022  06:48 PM           275,214 javaparser-1.0.8.jar
11/29/2022  06:48 PM           826,947 javassist-3.26.0-GA.jar
11/29/2022  06:48 PM            43,649 javax.activation-1.2.0.jar
11/29/2022  06:48 PM            67,231 javax.activation-api-1.2.0.jar
11/29/2022  06:48 PM            35,618 javax.annotation-api-1.3.jar
11/29/2022  06:48 PM            13,835 javax.inject-2.4.0-b31.jar
11/29/2022  06:48 PM           560,452 javax.mail-1.5.0.jar
11/29/2022  06:48 PM           133,628 javax.ws.rs-api-2.0.1.jar
11/29/2022  06:48 PM           145,331 jaxb-api-2.4.0-b180725.0427.jar
11/29/2022  06:48 PM         1,167,642 jaxb-runtime-2.4.0-b180725.0644.jar
04/18/2022  03:37 PM           862,588 jcifs-ng-2.1.3.jar
11/29/2022  06:48 PM           869,749 jcifs-ng-2.1.8.jar
11/29/2022  06:48 PM           180,439 jersey-client-2.22.1.jar
11/29/2022  06:48 PM           746,152 jersey-common-2.22.1.jar
11/29/2022  06:48 PM         1,046,673 jersey-guava-2.22.1.jar
11/29/2022  06:48 PM            79,243 jersey-media-multipart-2.22.1.jar
11/29/2022  06:48 PM            16,924 jersey-proxy-client-2.22.1.jar
09/02/2021  01:01 PM           580,207 jetty-util-9.4.21.v20190926.jar
11/29/2022  06:48 PM           623,868 jetty-util-9.4.44.v20210927.jar
11/29/2022  06:48 PM           703,015 joda-time-2.9.9.jar
11/29/2022  06:48 PM         1,151,049 language_support.jar
11/29/2022  06:48 PM           421,263 libphonenumber-8.12.50.jar
11/29/2022  06:48 PM           327,027 log4j-1.2-api-2.17.2.jar
11/29/2022  06:48 PM           326,308 log4j-api-2.17.2.jar
11/29/2022  06:48 PM         1,927,058 log4j-core-2.17.2.jar
11/29/2022  06:48 PM           447,668 looks-2.3.1.jar
11/29/2022  06:48 PM            76,633 miglayout-core-4.2.jar
11/29/2022  06:48 PM            22,520 miglayout-swing-4.2.jar
11/29/2022  06:48 PM            75,877 mimepull-1.9.7.jar
11/29/2022  06:48 PM           844,984 mirth-client-core.jar
11/29/2022  06:48 PM         2,984,061 mirth-client.jar
11/29/2022  06:48 PM         3,033,663 mirth-crypto.jar
11/29/2022  06:48 PM         5,633,794 mirth-vocab.jar
11/29/2022  06:48 PM         9,970,366 openjfx.jar
11/29/2022  06:48 PM           622,736 quartz-all-2.1.7.jar
11/29/2022  06:48 PM           146,649 reflections-0.9.10.jar
11/29/2022  06:48 PM           469,207 regions-2.15.28.jar
11/05/2020  05:57 PM           316,129 regions-2.9.14.jar
09/02/2021  01:01 PM         1,359,155 rhino-1.7.12.jar
11/29/2022  06:48 PM         1,377,387 rhino-1.7.13.jar
11/29/2022  06:48 PM         1,086,882 rsyntaxtextarea-2.5.6.jar
11/05/2020  05:57 PM            50,547 slf4j-api-1.7.28.jar
11/29/2022  06:48 PM            52,259 slf4j-api-1.7.30.jar
11/05/2020  05:57 PM            18,979 slf4j-log4j12-1.7.28.jar
11/29/2022  06:48 PM            20,328 slf4j-log4j12-1.7.30.jar
11/29/2022  06:48 PM           167,241 staxon-1.3.jar
11/29/2022  06:48 PM            45,931 swagger-annotations-2.0.10.jar
11/29/2022  06:48 PM         1,563,702 swingx-core-1.6.2.jar
11/29/2022  06:48 PM            77,179 userutil-sources.jar
11/29/2022  06:48 PM           124,317 utils-2.15.28.jar
11/05/2020  05:57 PM           101,276 utils-2.9.14.jar
11/29/2022  06:48 PM           556,450 velocity-engine-core-2.2.jar
11/29/2022  06:48 PM           230,042 velocity-tools-generic-3.0.jar
11/29/2022  06:48 PM           234,871 wizard.jar
11/29/2022  06:48 PM         1,333,223 xercesImpl-2.9.1.jar
11/29/2022  06:48 PM           133,330 xpp3-1.1.4c.jar
04/18/2022  03:37 PM           681,281 xstream-1.4.12.jar
11/29/2022  06:48 PM           693,019 xstream-1.4.19.jar
11/29/2022  06:48 PM           136,477 zip4j_1.3.3.jar
             106 File(s)     97,841,413 bytes
               2 Dir(s)  31,404,814,336 bytes free
```

**Reply by ab-mg-23** [reply]
You absolutely would want to remove the older files, including if there are dupes in the server-lib or extensions. I'm 99% sure the JVM loads in ascending order when a directory + wildcard is an element of the classpath, so the older versions of libs will get loaded not the latest.

**Comment by pacmano1** [comment]
I'd probably backup up the mirth install dir (e.g. c:\program files\Mirth... and wherever appdata is), completely remove mirth (and then delete the mirth install dir), then re-install and merge a small subset as files  as needed.   e.g. stuff in the old appdata folder (e.g. configuration.propeties, keystore.jks and server.id), and mirth.properties.  If there is currently stuff in custom-lib I would not restore that rather move it to Setting..Resorces being mindful of the far right columns.   This all assumes your mirth database is NOT Derby.  My intitial response was a concern over this line in mirth.properties `rhino.languageversion = es6`, but @ab-mg-23 's suggestion was far better.

**Comment by AaronKelley** [comment]
OK.  Here's what I did.

I did a new Mirth Connect 4.2.0 install on a different system.  Then, on the "problem" system, I shut down the Mirth service and moved out the folders "client-lib", "extensions", and "server-lib".  I replaced them with the same folders taken from the system where I did the new install (effectively cleaning out the duplicate/old libraries noted above).  Then I fired up the Mirth service again and confirmed that everything was up and running normally.

I then did a database backup and VM snapshot and attempted again the upgrade to 4.3.0.  All good this time!  I am thinking the upgrade to 4.5.0 will also be fine since I got hung up on the same error last time I tried that; I'll post back if that is not the case.

No idea why the installer did not clean up these old libraries on its own, or even which version of the installer made the mess (seems like it must have been messed up already when I was on 3.10), but now I know what to check for if something like this happens again.

I did not end up having to mess with anything in custom-lib.  I will look into moving those to Resources as @pacmano1 suggested.  Thanks to both of you for the feedback.

**Comment by AaronKelley** [comment]
Also thanks to @pacmano1 for [this snippet](https://github.com/pacmano1/Mirth-Snippets/blob/main/MirthSSH_4.5.md), right after upgrading to 4.5.0 I noticed that I have one stubborn SFTP server that I need to talk to that doesn't support anything "better" than ssh-rsa.
